sdks/java/harness/src/test/java/org/apache/beam/fn/harness/state/OrderedListUserStateTest.java

```diff
@@ -0,0 +1,684 @@
+              Caches.noop(),
+              fakeClient,
+              "instructionId",
+              createOrderedListStateKey("A"),
+              StringUtf8Coder.of());
+
+      assertThat(userState.read(), is(emptyIterable()));
+    }
+  }
+
+  @Test
+  public void testOperationsDuringNavigatingIterable() throws Exception {
+    FakeBeamFnStateClient fakeClient =
+        new FakeBeamFnStateClient(
+            timestampedValueCoder,
+            ImmutableMap.of(
+                createOrderedListStateKey("A", 1),
+                asList(A1, B1),
+                createOrderedListStateKey("A", 2),
+                asList(A2, B2),
+                createOrderedListStateKey("A", 3),
+                Collections.singletonList(A3),
+                createOrderedListStateKey("A", 4),
+                Collections.singletonList(A4)));
+
+    OrderedListUserState<String> userState =
+        new OrderedListUserState<>(
+            Caches.noop(),
+            fakeClient,
+            "instructionId",
+            createOrderedListStateKey("A"),
+            StringUtf8Coder.of());
+
+    Iterator<TimestampedValue<String>> iter = userState.read().iterator();
+    assertEquals(iter.next(), A1);
+
+    // Adding a C1 locally, but it should not be returned after B1 in the existing iterable.
+    userState.add(C1);
+    assertEquals(iter.next(), B1);
+    assertEquals(iter.next(), A2);
+
+    // Clearing range [2,4) locally, but B2 and A3 should still be returned.
+    userState.clearRange(Instant.ofEpochMilli(2), Instant.ofEpochMilli(4));
+    assertEquals(iter.next(), B2);
+    assertEquals(iter.next(), A3);
+
+    // Clearing all ranges locally, but A4 should still be returned.
+    userState.clear();
+    assertEquals(iter.next(), A4);
+  }
+
+  private ByteString encode(String... values) throws IOException {
+    ByteStringOutputStream out = new ByteStringOutputStream();
+    for (String value : values) {
+      StringUtf8Coder.of().encode(value, out);
+    }
+    return out.toByteString();
+  }
+
+  private StateKey createOrderedListStateKey(String key) throws IOException {
+    return StateKey.newBuilder()
+        .setOrderedListUserState(
+            StateKey.OrderedListUserState.newBuilder()
+                .setWindow(encode(encodedWindow))
+                .setTransformId(pTransformId)
+                .setUserStateId(stateId)
+                .setKey(encode(key)))
+        .build();
+  }
+
+  private StateKey createOrderedListStateKey(String key, long sortKey) throws IOException {
+    return StateKey.newBuilder()
+        .setOrderedListUserState(
+            StateKey.OrderedListUserState.newBuilder()
+                .setWindow(encode(encodedWindow))
+                .setTransformId(pTransformId)
+                .setUserStateId(stateId)
+                .setKey(encode(key))
+                .setRange(
+                    OrderedListRange.newBuilder().setStart(sortKey).setEnd(sortKey + 1).build()))
+        .build();
+  }
+}
```

Comment:
You've covered a lot of cases here, but this seems like the kind of thing where it'd be good to have fuzz tests too. Say you do dozens of random interspersed reads, writes, and clears. You could do the same thing on a naive List (as long as you kept your range small enough) and verify you got the same result. (Make sure you record a seed so you can reproduce failures.)

Reply:
OK. Please file an issue and assign it to yourself.

Reply:
Done. Opened a issue: https://github.com/apache/beam/issues/31592.

Thanks @robertwb!